### PR TITLE
asserts: introduce Add() and Find()

### PR DESCRIPTION
**Pull request by pedronis**
Add() in the final version keeps only the latest revision of an assertion.

Find() takes a full primary key so it returns either one or no assertion. Later will likely need something that probably only takes a prefix of the primary key and can do some further filtering, we need to look at flows/use cases.

### CODE REVIEW

File: asserts/database.go [hunk 1]
```diff
@@ -53,6 +57,12 @@ type DatabaseConfig struct {
 	TrustedKeys map[string][]PublicKey
 }
 
+// Well-known errors
+var (
+	ErrAssertionNotFound       = errors.New("assertion not found")
+	ErrAssertionNotSuperseding = errors.New("assertion does not supersede current one")
```
Comment (niemeyer) [inline asserts/database.go]: I suggest `s/Assertion//` here, because we're already in the context of the `asserts` package here (e.g. `asserts.ErrNotFound` seems about right).

File: asserts/database.go [hunk 2]
```diff
@@ -95,6 +107,41 @@ func (db *Database) atomicWriteEntry(data []byte, secret bool, subpath ...string
...
 }
 
+func (db *Database) removeEntry(subpath ...string) error {
+	fpath := filepath.Join(db.root, filepath.Join(subpath...))
+	return os.Remove(fpath)
+}
+
+func (db *Database) symlinkEntry(entry string, subpath ...string) error {
+	fpath := filepath.Join(db.root, filepath.Join(subpath...))
+	// TODO: move to helpers/helpers.go
```
Comment (niemeyer) [inline asserts/database.go]: I actually hope we can move the content in helpers to packages slightly more indicative of their function in the system. Names such as utils, helpers, tools, etc, tend to creep up for lack of clear focus.

File: asserts/database.go [hunk 3]
```diff
@@ -176,3 +223,101 @@ func (db *Database) Check(assert Assertion) error {
...
+		curRevStr, err := db.readlinkEntry(assertionsRoot, indexPath, "latest")
+		if err != nil {
+			return fmt.Errorf("broken assertion store, reading 'latest' revision symlink: %v", err)
+		}
+		curRev, err := strconv.Atoi(curRevStr)
+		if err != nil {
+			return fmt.Errorf("broken assertion store, extracting revision from 'latest' revision symlink: %v", err)
+		}
+		if curRev >= assert.Revision() {
+			return ErrAssertionNotSuperseding
```
Comment (niemeyer) [inline asserts/database.go]: I generally try to wait for a use case before making things a variable. Is there a clear reason why we'll want to test for this error, specifically, or is it just like the other errors in that we can just forward the issue to the user when necessary?

File: asserts/database.go [hunk 4]
```diff
@@ -176,3 +223,101 @@ func (db *Database) Check(assert Assertion) error {
...
+	primaryKey := make([]string, len(reg.primaryKey))
+	for i, k := range reg.primaryKey {
+		keyVal := assert.Header(k)
+		if keyVal == "" {
+			return fmt.Errorf("missing primary key header: %v", k)
+		}
+		// safety against '/' etc
+		primaryKey[i] = url.QueryEscape(keyVal)
+	}
+	indexPath := filepath.Join(string(assert.Type()), filepath.Join(primaryKey...))
```
Comment (niemeyer) [inline asserts/database.go]: Not sure about the nesting here.. it makes it convenient to search for items via their primary key order, but makes everything else a pain. That's a detail we can adjust soon, though.. please feel free to move this forward without this being addressed, and let's talk about it at the next opportunity.

File: asserts/database.go [hunk 5]
```diff
@@ -176,3 +223,101 @@ func (db *Database) Check(assert Assertion) error {
...
+	_, err = db.statEntry(assertionsRoot, indexPath)
+	switch {
+	case err == nil:
+		// directory for assertion is present
+		curRevStr, err := db.readlinkEntry(assertionsRoot, indexPath, "latest")
```
Comment (niemeyer) [inline asserts/database.go]: What would happen if there was both `{foo, latest, bar}` and `{foo}` as primary keys for different assertions?

File: asserts/database.go [hunk 6]
```diff
@@ -176,3 +223,101 @@ func (db *Database) Check(assert Assertion) error {
...
+		if keyVal == "" {
+			return fmt.Errorf("missing primary key header: %v", k)
+		}
+		// safety against '/' etc
+		primaryKey[i] = url.QueryEscape(keyVal)
+	}
+	indexPath := filepath.Join(string(assert.Type()), filepath.Join(primaryKey...))
+	_, err = db.statEntry(assertionsRoot, indexPath)
+	switch {
+	case err == nil:
```
Comment (niemeyer) [inline asserts/database.go]: The switch here for some reason seems to read a bit strangely. To help out, it feels like the logic inside this err == nil case stands on its own, so perhaps we can take it out into its own appropriately named function, and then have a shorter branches here.

What do you think?

File: asserts/database.go (hunk 5 above)
Comment (pedronis) [inline asserts/database.go]: @niemeyer we prefix the whole thing with the assertion type (line 247), and the current setup assume uniform full primary keys for assertions of the same type

File: asserts/database.go [hunk 7]
```diff
@@ -176,3 +223,101 @@ func (db *Database) Check(assert Assertion) error {
...
+		keyVal := headers[k]
+		if keyVal == "" {
+			return nil, fmt.Errorf("must provide primary key: %v", k)
+		}
+		primaryKey[i] = url.QueryEscape(keyVal)
+	}
+	indexPath := filepath.Join(string(assertionType), filepath.Join(primaryKey...))
+	_, err = db.statEntry(assertionsRoot, indexPath)
+	if err != nil {
+		if os.IsNotExist(err) {
```
Comment (niemeyer) [inline asserts/database.go]: This can be the first thing after statEntry, to avoid the double nesting (it can take nil).

File: asserts/database.go (hunk 5 above)
Comment (niemeyer) [inline asserts/database.go]: Either I understand, or I couldn't make the point clear. Let me try again: what happens if `"latest"` is used as part of a primary key? The fact you prefix it with a type doesn't change the issue.

Comment (pedronis) [inline asserts/database.go]: @niemeyer I don't understand, the path levels are fixed for a given assertion type

File: asserts/database.go (hunk 3 above)
Comment (pedronis) [inline asserts/database.go]: @niemeyer it seemed to me a higher level decision whether this is expected/unexpected, should be ignored... that's why I separated it out

File: asserts/database.go [hunk 8]
```diff
@@ -176,3 +223,110 @@ func (db *Database) Check(assert Assertion) error {
 	}
 	return fmt.Errorf("failed signature verification: %v", lastErr)
 }
+
+func (db *Database) checkAssertionRevisionAndClearLatest(indexPath string, newRev int) error {
```
Comment (niemeyer) [inline asserts/database.go]: Sorry for not being more clear. There's clear functionality here which is "get me the current revision of this assertion". This should be in a function on its own.. the clear latest and replace is very specific behavior to the adding of the new assertion, and can remain in the original function.

File: asserts/database.go [hunk 9]
```diff
@@ -176,3 +223,110 @@ func (db *Database) Check(assert Assertion) error {
...
+		return fmt.Errorf("broken assertion storage, reading 'latest' revision symlink: %v", err)
+	}
+	curRev, err := strconv.Atoi(curRevStr)
+	if err != nil {
+		return fmt.Errorf("broken assertion storage, extracting revision from 'latest' revision symlink: %v", err)
+	}
+	if curRev >= newRev {
+		return ErrNotSuperseding
+	}
+	err = db.removeEntry(assertionsRoot, indexPath, "latest")
```
Comment (niemeyer) [inline asserts/database.go]: This seems broken, though. We're not dealing with an atomic database, so we cannot afford to lose knowledge like this. If the system is reset while we are in the interim state, it will all break down.

Comment (pedronis) [inline asserts/database.go]: @niemeyer 'latest' is an optimisation, we can always recover by finding the top revision in the dir if it's not there, I didn't go there yet because as I said in my intro to the branch it wasn't clear to me if keeping old revision is something we want actually?

File: asserts/database.go [hunk 10]
```diff
@@ -176,3 +192,94 @@ func (db *Database) Check(assert Assertion) error {
...
+		return nil, fmt.Errorf("broken assertion storage, failed to read assertion: %v", err)
+	}
+	assert, err := Decode(encoded)
+	if err != nil {
+		return nil, fmt.Errorf("broken assertion storage, failed to decode assertion: %v", err)
+	}
+	return assert, nil
+}
+
+func (db *Database) readCurrentRevision(assertType AssertionType, primaryPath string) (int, error) {
```
Comment (niemeyer) [inline asserts/database.go]: Thanks, these are much nicer now.

File: asserts/database.go [hunk 11]
```diff
@@ -176,3 +192,94 @@ func (db *Database) Check(assert Assertion) error {
...
+	assert, err := db.readAssertion(assertType, primaryPath)
+	if err == ErrNotFound {
+		return -1, nil
```
Comment (niemeyer) [inline asserts/database.go]: That doesn't look right. We should never return -1 on err == nil. Was the intention to return 0 here?

Comment (niemeyer) [inline asserts/database.go]: (test missing, if so)

Comment (pedronis) [inline asserts/database.go]: @niemeyer if we return 0 adding the first revision (which is also 0) would fail, I'm marking not found with (-1, nil) but if you prefer to propagate an explicit NotFound and have bit more logic in the caller works for me

File: asserts/database.go [hunk 12]
```diff
@@ -176,3 +192,94 @@ func (db *Database) Check(assert Assertion) error {
...
+		// safety against '/' etc
+		primaryKey[i] = url.QueryEscape(keyVal)
+	}
+	primaryPath := filepath.Join(primaryKey...)
+	curRev, err := db.readCurrentRevision(assert.Type(), primaryPath)
+	if err != nil {
+		return err
+	}
+	if curRev >= assert.Revision() {
+		return ErrNotSuperseding
```
Comment (niemeyer) [inline asserts/database.go]: Can we have something explicit here until we need the exported variable:

    return fmt.Errorf("assertion added must be more recent than current one (adding %d, currently %d)", ...)

File: asserts/database.go (hunk 11 above)
Comment (niemeyer) [inline asserts/database.go]: Okay, so let's get rid of readCurrentRevision altogether, because it's really not adding much on top of readAssertion. More importantly, though: can we please establish this rule for future work: 

_Never return invalid data if err == nil._

Comment (pedronis) [inline asserts/database.go]: @niemeyer well I can proceed as you wish, and I agree about the rule but -1 wasn't invalid data as I was conceving the interface

**Comment by niemeyer** [comment]
LGTM, thank you!

File: asserts/database.go (hunk 11 above)
Comment (niemeyer) [inline asserts/database.go]: Erm, that's not sound. Let's talk online.

**Comment by chipaca** [comment]
LGTM :-)
